### PR TITLE
Refactor datavalues and datatypes

### LINKED ISSUE #810: Generic DataTypes with associate  ColumnChunk<Type>
**Summary**

Currently, it's hard to apply generic functions/classes  in `datafuse`, because we don't have our own types and arrays. Enum for types is not a good way in rust to implement generic structs.

I propose to introduce these ASAP. 

Related codes in other systems: 

https://github.com/pola-rs/polars/blob/master/polars/polars-core/src/datatypes.rs#L119-L155

Related issue:  
 #644

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary
Draft of #810 

###   Own *DataType* and *DFDataType*

- [x] DataType is an enum class which play kinds like arrow::DataType,  DataField and DataSchema depends on DataType
- [x] *DFDataTye*  is a trait DataType for datafuse, lots of DFDataTypeImpl can implement it, such as UInt8Type, UInt16Type.
    -  DFPrimitiveType, `DFPrimitiveType: ArrowPrimitiveType + DFDataType` which  stands for primitive types.
    -  DFNumericType,  `DFNumericType: DFPrimitiveType + ArrowNumericType {}` which stands for numeric types.
    -  DFFloatType, `DFFloatType:DFNumericType` stands for float types.
    - ....

- [x] own DataField and DataSchema based on new DataType, and implement convertor from/to ArrowTypes/ArrowSchema/ArrowField.
 

- [x] DataArray<T> where T: DFDataType

```
pub struct DataArray<T> {
    pub array: ArrayRef,
    phantom: PhantomData<T>,
}
```

- [x] Series   
- [x] migrate the ops and kernels for datavalues
- [x] migrate tests

## Changelog

- Improvement

## Related Issues

Fixes #810

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/datavalues/src/todos/data_array_arithmetic.rs [hunk 1]
```diff
@@ -8,55 +8,86 @@ use common_exception::ErrorCode;
 use common_exception::Result;
```
Comment (BohuTANG) [inline common/datavalues/src/todos/data_array_arithmetic.rs:8]: Really great @sundy-li 
What's the todos directory for?

File: common/aggregate_functions/src/aggregate_function.rs [hunk 2]
```diff
@@ -19,17 +19,18 @@ pub trait AggregateFunction: fmt::Display + Sync + Send + DynClone {
 
     fn as_any(&self) -> &dyn Any;
 
+    // TODO
```
Comment (zhang2014) [inline common/aggregate_functions/src/aggregate_function.rs:22]: What to do?

File: common/datablocks/src/kernels/data_block_scatter.rs [hunk 3]
```diff
@@ -2,53 +2,41 @@
...
-            let column = block.column(column_index);
-            let scattered_column = DataArrayScatter::scatter(&column, &indices, scatter_size)?;
-
-            for scattered_index in 0..scattered_column.len() {
-                scattered_columns[scattered_index * columns_size + column_index] =
-                    Some(scattered_column[scattered_index].clone());
-            }
+            let mut indices = array.into_no_null_iter();
+
+            let columns = unsafe {
```
Comment (zhang2014) [inline common/datablocks/src/kernels/data_block_scatter.rs:25]: Why need unsafe?

File: common/datavalues/src/arrays/arrow_array.rs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+            let array_data = ArrayData::builder(DataType::Boolean)
+                .len(self.len())
+                .offset(self.offset())
+                .add_buffer(invert)
+                .build();
+            BooleanArray::from(array_data)
+        }
+    }
+    fn is_not_null_mask(&self) -> BooleanArray {
```
Comment (zhang2014) [inline common/datavalues/src/arrays/arrow_array.rs:37]: add new line.

File: common/datavalues/src/arrays/comparison.rs [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+        Ok(impl_cmp_numeric_utf8_list!(self, rhs, series_equal))
+    }
+
+    fn neq(&self, rhs: &DFListArray) -> Result<DFBooleanArray> {
+        self.eq(rhs)?.not()
+    }
+
+    // following are not implemented because gt, lt comparison of series don't make sense
+    fn gt(&self, _rhs: &DFListArray) -> Result<DFBooleanArray> {
+        unimplemented!()
```
Comment (zhang2014) [inline common/datavalues/src/arrays/comparison.rs:521]: maybe Err(ErrorCode::Unimplemented(...)) is better?

**Comment by zhang2014** [comment]
LGTM

File: common/datablocks/src/kernels/data_block_scatter.rs (hunk 3 above)
Comment (sundy-li) [reply common/datablocks/src/kernels/data_block_scatter.rs:25]: scatter_unchecked is unsafe function, we don't need to check the index bound

**Review by BohuTANG — APPROVED** [review]
LGTM
Thank you @sundy-li

**Pull request merged**